x-pack/plugins/security_solution/server/endpoint/services/actions/utils.ts

```diff
@@ -0,0 +1,203 @@
+    expiration: actionRequest.expiration,
+    createdBy: actionRequest.user_id,
+    createdAt: actionRequest['@timestamp'],
+  };
+};
+
+interface ActionCompletionInfo {
+  isCompleted: boolean;
+  completedAt: undefined | string;
+}
+
+export const getActionCompletionInfo = (
+  /** List of agents that the action was sent to */
+  agentIds: string[],
+  /** List of action Log responses received for the action */
+  actionResponses: Array<ActivityLogActionResponse | EndpointActivityLogActionResponse>
+): ActionCompletionInfo => {
+  const completedInfo: ActionCompletionInfo = {
+    isCompleted: true,
+    completedAt: undefined,
+  };
+
+  const responsesByAgentId = mapActionResponsesByAgentId(actionResponses);
+
+  for (const agentId of agentIds) {
+    if (!responsesByAgentId[agentId] || !responsesByAgentId[agentId].isCompleted) {
+      completedInfo.isCompleted = false;
+      break;
+    }
+  }
+
+  // If completed, then get the completed at date
+  if (completedInfo.isCompleted) {
+    for (const normalizedAgentResponse of Object.values(responsesByAgentId)) {
+      if (
+        !completedInfo.completedAt ||
+        completedInfo.completedAt < (normalizedAgentResponse.completedAt ?? '')
+      ) {
+        completedInfo.completedAt = normalizedAgentResponse.completedAt;
+      }
+    }
+  }
+
+  return completedInfo;
+};
+
+interface NormalizedAgentActionResponse {
+  isCompleted: boolean;
+  completedAt: undefined | string;
+  fleetResponse: undefined | ActivityLogActionResponse;
+  endpointResponse: undefined | EndpointActivityLogActionResponse;
+}
+
+type ActionResponseByAgentId = Record<string, NormalizedAgentActionResponse>;
+
+/**
+ * Given a list of Action Responses, it will return a Map where keys are the Agent ID and
+ * value is a object having information about the action responsess associated with that agent id
+ * @param actionResponses
+ */
+const mapActionResponsesByAgentId = (
+  actionResponses: Array<ActivityLogActionResponse | EndpointActivityLogActionResponse>
+): ActionResponseByAgentId => {
+  const response: ActionResponseByAgentId = {};
+
+  for (const actionResponse of actionResponses) {
+    if (actionResponse.type === 'fleetResponse' || actionResponse.type === 'response') {
+      const agentId = getAgentIdFromActionResponse(actionResponse);
+
+      if (!response[agentId]) {
+        response[agentId] = {
+          isCompleted: false,
+          completedAt: undefined,
+          fleetResponse: undefined,
+          endpointResponse: undefined,
+        };
+      }
+
+      if (actionResponse.type === 'fleetResponse') {
+        response[agentId].fleetResponse = actionResponse;
+      } else {
+        response[agentId].endpointResponse = actionResponse;
+      }
+
+      response[agentId].isCompleted = Boolean(response[agentId].endpointResponse);
+
+      if (response[agentId].isCompleted) {
+        response[agentId].completedAt = response[agentId].endpointResponse?.item.data['@timestamp'];
```

Comment:
Note here: I'm using the Event's `@timestamp` here instead of the `completed_at`:

https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/endpoint/types/actions.ts#L89-L103


This should be ok, right?